File: design-patterns/JavaDesignPatterns/src/com/bigboxcode/designpattern/builder/vechicle/CarBuilder.java
package com.bigboxcode.designpattern.builder.vechicle;

public class CarBuilder implements VehicleBuilder {
    private int wheel;
    private int engine;
    private int seat;
    private boolean interior;
    private int door;

    @Override
    public void addWheel(int noOfWheel) {
        System.out.println("Add " + noOfWheel + " wheels");

        this.wheel += noOfWheel;
    }

    @Override
    public void addEngine(int noOfEngine) {
        System.out.println("Add " + noOfEngine + " engine");

        this.engine += noOfEngine;
    }

    @Override
    public void addSeat(int noOfSeat) {
        System.out.println("Add " + noOfSeat + " Seat");

        this.seat = noOfSeat;
    }

    @Override
    public void addInterior() {
        System.out.println("Add interior");

        this.interior = true;
    }

    @Override
    public void addDoor(int noOfDoor) {
        System.out.println("Add " + noOfDoor + " door");

        this.door += noOfDoor;
    }

    @Override
    public void addWing(int noOfWings) throws Exception {
        throw new Exception("Can not add wings");
    }

    public Car build() {
        Car car = new Car(wheel, engine, seat, door, interior);

        return car;
    }

}

File: design-patterns/php/builder/vehicle/CarBuilder.php
<?php
// CarBuilder.php

namespace BigBoxCode\DesignPattern\Builder\Vehicle;


class CarBuilder implements VehicleBuilder {
    private int $wheel = 0;
    private int $engine = 0;
    private int $seat = 0;
    private bool $interior;
    private int $door = 0;

    public function addWheel(int $noOfWheel): void {
        echo "Add " . $noOfWheel . " wheels\n";

        $this->wheel += $noOfWheel;
    }

    public function addEngine(int $noOfEngine): void {
        echo "Add " . $noOfEngine . " engine\n";

        $this->engine += $noOfEngine;
    }

    public function addSeat(int $noOfSeat): void {
        echo "Add " . $noOfSeat . " Seat\n";

        $this->seat = $noOfSeat;
    }

    public function addInterior(): void {
        echo "Add interior\n";

        $this->interior = true;
    }

    public function addDoor(int $noOfDoor): void {
        echo "Add " . $noOfDoor . " door\n";

        $this->door += $noOfDoor;
    }

    public function addWing(int $noOfWing): void {
        throw new \Exception("Can not add wings");
    }

    public function build(): Car {
        return new Car($this->wheel, $this->engine, $this->seat, $this->door, $this->interior);
    }
}